README
The Mobile Web Project was initially developed during the summer of 2009 and is a fork of the original v0.9 release of the MIT Mobile Web project that can still be found on SourceForge. The project was updated in May 2010. It is a product designed to make it easier for higher education institutions to deliver mobile-optimized information and services. It helps deliver task-based content like maps, events, and directory information optimized for device "families." This project is still under active development.

Full documentation can be found at http://mobilewebosp.pbworks.com/
Updates to the project can be found on Twitter at http://twitter.com/mobilewebosp/


CHANGELOG
THIS CHANGELOG IS AN ATTEMPT TO DOCUMENT CHANGES TO THIS PROJECT.
SUPER MINOR REVISION NUMBERS WON'T SHOW UP IN GITHUB BUT IT'LL HELP
ME ORGANIZE WHEN I MADE A SET OF FIXES

MOBI-v2.5.0 (IN PROGRESS)
 - THIS IS SIMPLY AN UPDATE TO KEEP TRACK OF MY PROGRESS
 - FIX: device detection is now it's own class
 - FIX: internal stats have been tweaked to make it easier to add new sections & more granular reports on devices
 - FIX: google analytics now works with jqtouch better
 - FIX: weather section has been updated to work with the 2.x file structure
 - FIX: <?php tags have been added to the beginning of php files per php.ini recommendation
 - FIX: example vhost has been updated to properly restrict access to sensitive files
 - FIX: main about page is now a .php file so the GA code can be easily included
 - FIX: the "full site" icon for the default view is now correct
 - FIX: the rss lib will now allow articles with the same title (deprecated with addition of simple-rss)
 - FIX: when no results are returned from the map search a "no results" message is shown
 - FIX: requirements page has been updated to reflect 2010 devices
 - FIX: small bug with displaying dates in calendar favorites
 - FIX: upgraded map basic templates to use Google Static Maps v2 including custom icons based on WebKit template markers
 - FIX: upgraded map touch templates to use Google Maps v3 so that no API key is required
 - FIX: removed need to configure map markers. now requires markers to match name of types & subtypes in database.
 - ADD: mysql support has been added via support for MDB2. other databases like postgres and mssql should also be supported
 - ADD: a quick set-up system. refer to the INSTALL file
 - ADD: info.yml files for each section to provide basic information
 - ADD: rss feeds cached thanks to simple-rss support
 - ADD: an include file in emergency for showing the same emergency message across all three templates
 - ADD: better accessibility support for the webkit templates by utilizing ARIA roles & focus() on pageAnimationEnd
 - ADD: device detection api so 3rd party campus services can use mobi as a central mobile classification system
 - ADD: minifier now automatically discovers javascript to include so it doesn't require configuration
 - ADD: minifier can cache automatically discovered javascript for performance reasons
 - ADD: general adapter class for figuring out if modules support adapters
 - ADD: calendar adapter class so the calendar module is easier to customize from a data source perspective
 - ADD: map adapter class so the map module is easier to customize from a data source perspective
 - ADD: general federated search support for displaying search results from multiple modules in one view
 - ADD: calendar federated search support
 - ADD: map federated search support
 - ADD: the local map adapter now allows for a unique identifier field for matching up locations to their parent locations more easily
 - ADD: "deep linking" to content is now supported. currently can only return to 'home' instead of section home for deep, detail links.
 - THX: thanks to robin8 for all the great feedback & diffs for the fixes
 - THX: thanks to dlmulder for the feedback on the accessibility support

todo:
 - email iCal event for calendar
 - email link to event
 - just a greater "share" panel or something? like tweet and fb or something? seems ugly
 - add 'section' link support to deep linking
 - test the hell out deep linking
 - need to clean-up webkit view of federated search results
 - need to add content adapter support to people
 - need to add federated search support to people
 - dining services section
 - location services
 - google search appliance module
 - the google bubble for adding item to homescreen

MOBI-v2.0.2 (August 22, 2010)
 - FIX: the google analytics tracking graphic on the basic templates has been fixed. also addresses an issue where
        this caused a double-hit to be recorded in the Pageviews table
 - FIX: default theme homescreen icons have been updated
 - ADD: default news icon
 - THX: thanks to kto for the heads up on the icons in the default theme

MOBI-v2.0.1 (August 13, 2010)
 - ADD: proper detection for opera mobile
 - FIX: device detection upgraded to support PHP 5.3. removed references to eregi()
 - FIX: input color in the default theme is now black
 - FIX: $theme var in web/templates/touch/form.html now properly populates
 - FIX: category dropdown in maps now properly refreshes results by selected category
 - THX: thanks to robin8 for all the great feedback & diffs for the fixes

MOBI-v2.0.0 (May 14, 2010)
 - ADD: theme support
 - ADD: jQuery & jQTouch support for interactive effects in webkit templates
 - ADD: basic templates now support mobile-optimized Google Analytics
 - ADD: "favorites" feature in calendar, people, and map
 - ADD: using 'min' library to compress JS & CSS on the fly
 - ADD: support for JSON export of data to help support the iWVU iPhone app codebase
 - FIX: now supports basic, touch, and webkit templates
 - FIX: cleaned up base code and reorganized files
 - FIX: updated the example virtual host config to support compression of files

MOBI-v1.0.1 (November 28 2009)
 - FIX: added global contact add to send_email() of web/people/vcard.php
 - FIX: removed a superfluous variable from line #111 of web/people/vcard.php
 - FIX: added the ability to set var to allow mobile-friendly map tiles for static maps
 - FIX: fixed a syntax error in web/people/detail.php on line #20
 - FIX: added isset($query) to web/map/sp/index.html to fix unset variable error
 - FIX: added $city_state to config.gen.copy.inc.php so that the map results show the proper city/state


lib/stellar.php
<?php
/**
 * Copyright (c) 2008 Massachusetts Institute of Technology
 * 
 * Licensed under the MIT License
 * Redistributions of files must retain the above copyright notice.
 * 
 */

require_once "db.php";

class StellarData {
  private static $courses = array(
    '1'    => 'Civil and Environmental Engineering',
    '2'    => 'Mechanical Engineering',
    '3'    => 'Materials Science and Engineering',
    '4'    => 'Architecture',
    '5'    => 'Chemistry',
    '6'    => 'Electrical Engineering and Computer Science',
    '7'    => 'Biology',
    '8'    => 'Physics',
    '9'    => 'Brain and Cognitive Sciences',
    '10'   => 'Chemical Engineering',
    '11'   => 'Urban Studies and Planning',
    '12'   => 'Earth, Atmospheric, and Planetary Sciences',
    '13'   => 'Ocean Engineering',
    '14'   => 'Economics',
    '15'   => 'Management',
    '16'   => 'Aeronautics and Astronautics',
    '17'   => 'Political Science',
    '18'   => 'Mathematics',
    '20'   => 'Biological Engineering',
    '21'   => 'Humanities',
    '21A'  => 'Anthropology',
    '21F'  => 'Foreign Languages and Literatures',
    '21H'  => 'History',
    '21L'  => 'Literature',
    '21M'  => 'Music and Theater Arts',
    '21W'  => 'Writing and Humanistic Studies',
    '22'   => 'Nuclear Science and Engineering',
    '24'   => 'Linguistics and Philosophy',
    'CMS'  => 'Comparative Media Studies',
    'CSB'  => 'Computational and Systems Biology',
    'ESD'  => 'Engineering Systems Division',
    'HST'  => 'Health Sciences and Technology',
    'MAS'  => 'Media Arts and Sciences',
    'SP'   => 'Special Programs',
    'STS'  => 'Science, Technology, and Society',
  );

  private static $not_courses = array("SP");

  private static $base_url = "http://stellar.mit.edu/courseguide/course/";

  private static $rss_url = "http://stellar.mit.edu/SRSS/rss/course/";

  public static function get_courses() {
    $out = array();
    foreach(self::$courses as $id => $name) {
      $out[$id] = array(
        "name"      => $name,
        "is_course" => !in_array($id, self::$not_courses)
      );
    }
    return $out;
  }

  public static function get_course($id) {
    return array(
      "name"      => self::$courses[$id],
      "is_course" => !in_array($id, self::$not_courses)
    );
  }

  public static function get_others() {
    $all_courses = self::get_courses();
    $out = array();
    foreach($all_courses as $id => $course) {
      if(!preg_match("/^\d/", $id)) {
        $out[$id] = $course;
      }
    }
    return $out;
  }

  private static function getTag($xml_obj, $tag) {
    $list = $xml_obj->getElementsByTagName($tag);
    if($list->length == 0) {
      throw new Exception("no elements of type $tag found");
    }
    if($list->length > 1) {
      throw new Exception("elements of type $tag not unique, {$list->length} found");
    }
    return $list->item(0);
  }

  private static function getTagVal($xml_obj, $tag) {
    return self::getTag($xml_obj, $tag)->nodeValue;
  }

  private static function getTagVals($xml_obj, $tag) {
    $nodes = $xml_obj->getElementsByTagName($tag);
    $vals = array();
    foreach($nodes as $node) {
      $vals[] = $node->nodeValue;
    }
    return $vals;
  }

  public static function get_term_data() {
    $month = (int) date('m');
    return array(
      "year" => date('y'),
      "season" => ($month <= 7) ? 'sp' : 'fa'
    );
  }

  public static function get_term() {
    $data = self::get_term_data();
    return $data["season"] . $data["year"];
  }

  public static function get_term_text() {
    $data = self::get_term_data();
    $seasons = array("sp" => "Spring", "fa" => "Fall");    
    return $seasons[ $data["season"] ] . " 20" . $data["year"];
  }

  public static function get_classes($course) {
    if(!in_array($course, array_keys(self::$courses))) {
      throw new Exception("$course not a valid course ID-number");
    }

    $term = self::get_term();
    $xml_obj = new DOMDocument();

    error_reporting($error_reporting & ~E_WARNING);
      $xml = file_get_contents(self::$base_url . "$course/$term/index.xml");
    error_reporting($error_reporting);

    if($xml == "") {
      // if failed to grab xml feed, then run the generic error handler
      throw new DataServerException(self::$base_url . "$course/$term/index.xml is experiencing problems");
    }

    $xml_obj->loadXML($xml);

    $root = $xml_obj->documentElement;
    $body = self::getTag($xml_obj, 'courseTerm');
    $list = $body->getElementsByTagName('term');
    $subjects = $list->item(0);
    $classes_xml = self::getTag($body, 'subjects');
    $classes = $classes_xml->getElementsByTagName('subject');

    $stellar_classes = array();

    foreach($classes as $class_xml) {
      // the code currently assumes a class has one stellarSite
      // not sure how to handle the more general case
      $stellarSites = $class_xml->getElementsByTagName('stellarSites');
      if($stellarSites->length == 0) {
        // no stellar site for class so just continue onto the next class
        continue;
      } 
       
      $stellarSite = $stellarSites->item(0)->getElementsByTagName('stellarSite')->item(0);
      $stellarURL = self::getTag($stellarSite, 'stellarUrl')->nodeValue;
      preg_match('/\/(\w*\.\w*)$/', $stellarURL, $match);
      

      // extract times and locations
      $times_xml = self::getTag($stellarSite, 'times');
      $items = $times_xml->getElementsByTagName('item');
      $times = array();
      foreach($items as $item) {
        $times[] = array(
	  "title" => self::getTagVal($item, 'title'),
	  "time" => self::getTagVal($item, 'time'),
	  "location" => self::getTagVal($item, 'location'),
	);
      }

      // extract the class staff
      $staff_xml = self::getTag($stellarSite, 'staff');
      $staff = array(
        'instructors' => self::getStaff($staff_xml, 'instructors'),
        'tas' => self::getStaff($staff_xml, 'tas'),
      );

      $stellar_classes[] = array(
        "masterId" => self::getTagVal($class_xml, 'masterSubjectId'),
        "rssId" => $match[0],
        "id" => self::getTagVal($class_xml, 'subjectId'),
        "name" => self::getTagVal($class_xml, 'name'),
        "title" => $class_xml->getElementsByTagName('title')->item(0)->nodeValue,
        "description" => self::getTagVal($class_xml, 'description'),
        "times" => $times,
        "staff" => $staff,
      );
    }
    return $stellar_classes;
  }

  private static function getStaff($staff_xml, $type) {
    $child = $staff_xml->getElementsByTagName($type);
    if($child->length == 1) {
      return self::getTagVals($child->item(0), 'fullName');
    } else {
      return array();
    }
  }

  public static function get_class_id($id) {
    $db = db::$connection;

    $parts = explode('.', trim($id));
    $course_id = $parts[0];
    $class_id = $parts[1];

    $sql = "SELECT main_course_id, main_class_id FROM ClassID WHERE (this_course_id=? AND this_class_id=?)";

    $stmt = $db->prepare($sql);
    $stmt->bind_param('ss', $course_id, $class_id);
    $stmt->bind_result($result_course_id, $result_class_id);
    $stmt->execute();
    if($stmt->fetch()) {
      $result = "{$result_course_id}.{$result_class_id}";
    }
    $stmt->free_result();
    if(!$result) {
      $result = $id;
    }
    return $result;
  }

  public static function get_class_info($id) {
    $parts = explode('.', $id);
    $course_id = $parts[0];

    $classes = self::get_classes($course_id);

    foreach($classes as $aClass) {
      if($aClass['masterId'] == $id) {
        $class = $aClass;
        break;
      }
    }
   
    //get the annoucements
    $rss_id = $class['rssId'];
    $rss_obj = new DOMDocument();
    $term = self::get_term();
    $rss = file_get_contents(self::$rss_url . "$course_id/$term/$rss_id/");  
    $rss_obj->loadXML($rss);    
    $rss_root = $rss_obj->documentElement;

    $class["announcements"] = array();
    foreach($rss_root->getElementsByTagName('item') as $item) {
      $title = self::getTag($item, 'title')->nodeValue;
      $colon_pos = strpos($title, ":");
      $class["announcements"][] = array(
   	  "date"     => date_parse(self::getTag($item, 'pubDate')->nodeValue),
          "unixtime" => strtotime(self::getTag($item, 'pubDate')->nodeValue),
          "title"    => substr($title, $colon_pos + 1),
          "text"     => self::getTag($item, 'description')->nodeValue
      );
    }
    return $class;
  }

  public static function search_classes($terms) {
    $db = db::$connection;
 
    //since courses numbers are uppercase
    //convert to uppercase
    $terms = strtoupper($terms);

    $words = split(' ', $terms);
    $ids = array();
    $new_words = array();
    
    foreach($words as $word) {
      preg_match('/^(\w*)(.*)/', $word, $match);
      if(in_array($match[1], array_keys(self::$courses))) {
        if(preg_match('/^\..*/', $match[2])) {
          $ids[] = array($match[1], substr($match[2], 1));
        } else {
          $ids[] = array($match[0]);
        }
      } elseif($word) {
        $new_words[] = $word;
      }
    }
    $new_words = implode(' ', $new_words);

    $sql = "SELECT DISTINCT Class.* FROM Class, ClassID WHERE (ClassID.main_course_id=Class.course_id AND ClassID.main_class_id=Class.class_id)";
    
    // do an OR search with ClassID's
    $or_sqls = array();
    foreach($ids as $id) {
      if(count($id) == 1) {
        $or_sqls[] = "(ClassID.this_course_id = '{$id[0]}' OR Class.name like '% {$id[0]}.%' OR Class.name like '{$id[0]}.%')";
      }	else {
        $or_sqls[] = "((ClassID.this_course_id = '{$id[0]}' AND ClassID.this_class_id like '{$id[1]}%') OR Class.name like '% {$id[0]}.{$id[1]} %' OR Class.name like '{$id[0]}.{$id[1]}%')";
      }
    }

    if(count($or_sqls) > 0) {
      $sql .= "AND (" . implode(" OR ", $or_sqls) . ")";
    }

    if($new_words) {
      $sql .= "AND (Class.title like '%" . $db->escape_string($new_words) . "%')";
    }

    $stmt1 = $db->prepare($sql);
    $stmt1->bind_result($course_id, $class_id, $title, $name);
    $stmt1->execute();
    $classes = array();
    while($stmt1->fetch()) {
      $classes[] = array(
        "masterId"    => $course_id . "." . $class_id,
        "title" => $title,
        "name"   => $name,
      );
    }
    $stmt1->free_result();

    $stmt2 = $db->prepare(
      "SELECT this_course_id, this_class_id FROM ClassID WHERE main_course_id=? AND main_class_id=?"
    );

    foreach($classes as $index => $class) {
      //find all the alternate IDs for $class
      $stmt2->bind_param('ss', $class['ids'][0], $class['ids'][1]);
      $stmt2->bind_result($course_id, $class_id);
      $stmt2->execute();

      //store the alternate ids     
      $classes[$index]['ids'] = array();
      while($stmt2->fetch()) {
        //add each alternate id
        $classes[$index]['ids'][] = $course_id . "." . $class_id;
      }
      $stmt2->free_result();
    }
    return $classes;        
  }


  public static function populate_db() {
    $db = db::$connection;
    $stmt1 = $db->prepare( "INSERT INTO Class (course_id, class_id, title, name) values (?, ?, ?, ?)" );

    $stmt2 = $db->prepare( "SELECT COUNT(*) FROM Class WHERE course_id=? AND class_id=?" );

    $stmt3 = $db->prepare( "INSERT INTO ClassID (main_course_id, main_class_id, this_course_id, this_class_id) values (?, ?, ?, ?)" );

    foreach(self::$courses as $course_id => $course_title) {
      foreach(self::get_classes($course_id) as $class) {
        $id_parts = explode('.', $class['masterId']);

        $stmt2->bind_param('ss', $id_parts[0], $id_parts[1]);
        $stmt2->bind_result($count);
        $stmt2->execute();
        $stmt2->fetch();
        $stmt2->free_result();

        // only add the the table if no pre-existing entry exists
        if($count == 0) {
          $stmt1->bind_param('ssss', $id_parts[0], $id_parts[1], $class['title'], $class['name']); 
          $stmt1->execute();
        }

        $this_id_parts = explode('.', $class['id']);
        $stmt3->bind_param('ssss', $id_parts[0], $id_parts[1], $this_id_parts[0], $this_id_parts[1]);
        $stmt3->execute();
      }
    }
  }
}  
?>
